contracts/utils/Base58.sol

```diff
@@ -0,0 +1,184 @@
+    bytes internal constant _LOOKUP_TABLE =
+        hex"000102030405060708ffffffffffffff090a0b0c0d0e0f10ff1112131415ff161718191a1b1c1d1e1f20ffffffffffff2122232425262728292a2bff2c2d2e2f30313233343536373839";
+
+    /**
+     * @dev Encode a `bytes` buffer as a Base58 `string`.
+     */
+    function encode(bytes memory data) internal pure returns (string memory) {
+        return string(_encode(data));
+    }
+
+    /**
+     * @dev Decode a Base58 `string` into a `bytes` buffer.
+     */
+    function decode(string memory data) internal pure returns (bytes memory) {
+        return _decode(bytes(data));
+    }
+
+    function _encode(bytes memory data) private pure returns (bytes memory encoded) {
+        // For reference, solidity implementation
+        // unchecked {
+        //     uint256 dataLeadingZeros = data.countLeading(0x00);
+        //     uint256 length = dataLeadingZeros + ((data.length - dataLeadingZeros) * 8351) / 6115 + 1;
+        //     encoded = new bytes(length);
+        //     uint256 end = length;
+        //     for (uint256 i = 0; i < data.length; ++i) {
+        //         uint256 ptr = length;
+        //         for (uint256 carry = uint8(data[i]); ptr > end || carry != 0; --ptr) {
+        //             carry += 256 * uint8(encoded[ptr - 1]);
+        //             encoded[ptr - 1] = bytes1(uint8(carry % 58));
+        //             carry /= 58;
+        //         }
+        //         end = ptr;
+        //     }
+        //     uint256 encodedCLZ = encoded.countLeading(0x00);
+        //     length -= encodedCLZ - dataLeadingZeros;
+        //     encoded.splice(encodedCLZ - dataLeadingZeros);
+        //     for (uint256 i = 0; i < length; ++i) {
+        //         encoded[i] = _TABLE[uint8(encoded[i])];
+        //     }
+        // }
+
+        // Assembly is ~50% cheaper for buffers of size 32.
+        assembly ("memory-safe") {
+            function clzBytes(ptr, length) -> i {
+                for {
+                    i := 0
+                } and(iszero(byte(0, mload(add(ptr, i)))), lt(i, length)) {
+                    i := add(i, 1)
+                } {}
+            }
+
+            encoded := mload(0x40)
+            let dataLength := mload(data)
+
+            // Count number of zero bytes at the beginning of `data`. These are encoded using the same number of '1's
+            // at then beginning of the encoded string.
+            let dataLeadingZeros := clzBytes(add(data, 0x20), dataLength)
+
+            // Initial encoding length: 100% of zero bytes (zero prefix) + 138% of non zero bytes + 1
+            let slotLength := add(add(div(mul(sub(dataLength, dataLeadingZeros), 138), 100), dataLeadingZeros), 1)
+
+            // Zero the encoded buffer
+            for {
+                let i := 0
+            } lt(i, slotLength) {
+                i := add(i, 0x20)
+            } {
+                mstore(add(add(encoded, 0x20), i), 0)
+            }
```

Comment:
```suggestion
            calldatacopy(add(encoded, 0x20), calldatasize(), slotLength)
```

Reply:
solady's hacks making it into OZ, one at a time